### PR TITLE
Adding Yahoo adapter for 1.0.1

### PR DESCRIPTION
**Pull request by nimrich**

### CODE REVIEW

File: ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    if (android.os.Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN) {
+      Log.e(TAG, "Yahoo Ads SDK minimum supported API is 16");
+
+      return false;
+    }
+
+    boolean success = true;
+    if (!YASAds.isInitialized()) {
+      if (!(context instanceof Activity)) {
+        Log.e(TAG, "StandardEdition.initialize must be explicitly called with an Activity" +
```
Comment (joshuajplagonera) [inline ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java:253]: Should `StandardEdition.initialize` be renamed?

File: ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+      }
+      try {
+        Application application = ((Activity) context).getApplication();
+        Log.d(TAG, "Initializing using site ID: " + siteId);
+        success = YASAds.initialize(application, siteId);
+      } catch (Exception e) {
+        Log.e(TAG, "Error occurred initializing Yahoo Ads SDK, ", e);
+
+        return false;
+      }
```
Comment (joshuajplagonera) [inline ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java:272]: If `YASAds` requires an `Activity` context to initialize the SDK, is there any reason why we are casting it to an `Application` context here?

**Review by joshuajplagonera — CHANGES_REQUESTED** [review]

File: ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java (hunk 2 above)
Comment (nimrich) [reply ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java:272]: The actual initialize call here needs an application context.  However, this call further down:
`YASAds.getActivityStateManager().setState((Activity) context,
        ActivityStateManager.ActivityState.RESUMED);`
requires an Activity and is also needed for initialization.  

Hopefully, that makes sense!

File: ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java (hunk 1 above)
Comment (nimrich) [reply ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java]: Yes, good catch!

**Comment by nimrich** [comment]
Addressed feedback.  Thanks!

File: ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+
+  @Override
+  public void onResume() {
+
+  }
+
+  /**
+   * Checks whether Yahoo SDK is initialized, if not initializes Yahoo SDK.
+   */
+  protected static boolean initializeSDK(final Context context, final String siteId) {
```
Comment (joshuajplagonera) [inline ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java:242]: We might refactor this method so we may better catch the errors in this method (Context, missing Site ID, etc).

File: ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java (hunk 2 above)
Comment (joshuajplagonera) [reply ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java:272]: Thank you for the explanation.
Can you add a comment here stating this behavior for future reference?

File: ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapterUtils.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+import com.yahoo.ads.RequestMetadata;
+import com.yahoo.ads.YASAds;
+import java.util.ArrayList;
+
+class YahooAdapterUtils {
+
+  private static final String PLACEMENT_KEY = "placement_id";
+  private static final String ORANGE_PLACEMENT_KEY = "position";
+  private static final String DCN_KEY = "dcn";
+  private static final String MEDIATOR_ID = "AdMobVAS-" + BuildConfig.VERSION_NAME;
```
Comment (joshuajplagonera) [inline ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapterUtils.java:23]: Just to confirm, this `VERSION_NAME` is the Yahoo Ads version name, or the adapter version name?

File: ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooBannerRenderer.java [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+  public void render(@NonNull final Context context, MediationBannerListener listener,
+      Bundle serverParameters, AdSize adSize, MediationAdRequest mediationAdRequest,
+      Bundle mediationExtras) {
+    bannerListener = listener;
+    String siteId = YahooAdapterUtils.getSiteId(serverParameters, mediationExtras);
+    MediationBannerAdapter adapter = bannerAdapterWeakRef.get();
+    if (TextUtils.isEmpty(siteId)) {
+      Log.e(TAG, "Failed to request ad: siteID is null or empty.");
+      if (bannerListener != null && adapter != null) {
+        bannerListener.onAdFailedToLoad(adapter, AdRequest.ERROR_CODE_INVALID_REQUEST);
```
Comment (joshuajplagonera) [inline ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooBannerRenderer.java:67]: Adding onto the previous comment, we will update this to support the new `AdError` API.

File: ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java (hunk 2 above)
Comment (nimrich) [reply ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java]: Added a comment regarding this.

File: ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapterUtils.java (hunk 4 above)
Comment (nimrich) [reply ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapterUtils.java]: This should be the adapter version name.  I did need to correct this to, "AdMobYAS."

File: ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooBannerRenderer.java (hunk 5 above)
Comment (nimrich) [reply ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooBannerRenderer.java]: Updated the onAdFailedToLoad calls to pass an AdError.

File: ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java (hunk 3 above)
Comment (nimrich) [reply ThirdPartyAdapters/yahoo/yahoo/src/main/java/com/google/ads/mediation/yahoo/YahooAdapter.java:247]: We do have checks for the Context and Site ID.  Could you provide a little more detail on what you had in mind.  Much thanks!

**Comment by chrisgiglio** [comment]
Please rebase this PR against the new `main` branch

**Comment by nimrich** [comment]
@chrisgiglio we are actually about to release the SDK rebranded as the Yahoo Mobile SDK.  I'll reopen this PR (or create a new one) once we have completed the release.  Thanks for the help!

**Pull request closed without merging**